### ISSUE TITLE
Zero-width space gets rendered as [obj]

### ISSUE THREAD

**Issue by rspilker** [DavBfr/dart_pdf#915]
**Describe the bug**
If a text contains a zero-width space `\u200b`, the output renders a dashed box with the text `obj` in it that overlaps the next characters. 

I've opened the document in both Chrome and Adobe Acrobat Reader DC, and both display this behavior.

I'm using the OpenSans font, did not test any other fonts. However, I assume that OpenSans is used so much that that alone warrants some solution.

As far as I understand, if the font file would not contain the offending character, I would have gotten an exception.

**To Reproduce**
Code snippet to reproduce the behavior:
```dart
import 'dart:io';
import 'package:pdf/widgets.dart';

Future<void> main() async {
  final openSans = Font.ttf(File('reproduction/OpenSans-Regular.ttf')
      .readAsBytesSync()
      .buffer
      .asByteData());

  final pdf = Document(theme: ThemeData.withFont(base: openSans));

  final content = 'Hello \u200b     \u200bWorld!';
  pdf.addPage(Page(build: (Context context) => Text(content)));

  File('reproduction/zerowidthspace.pdf').writeAsBytesSync(await pdf.save());
}
```

**Expected behavior**
I expected the zero-with space not to be rendering anything.

**Screenshots**
![image](https://user-images.githubusercontent.com/93318/149601979-1a9e55a6-45a3-4acd-951e-9be3a7a795bb.png)

[zerowidthspace.pdf](https://github.com/DavBfr/dart_pdf/files/7873949/zerowidthspace.pdf)

**Flutter Doctor**
Plain dart

```
```

**Desktop (please complete the following information):**
 - [ ] iOS
 - [ ] Android
 - [ ] Browser
 - [x] Windows
 - [ ] Linux

**Smartphone (please complete the following information):**
No phone

**Additional context**
Workaround: remove all zero-with spaces from any text before adding it to the pdf.

**Comment by rspilker** [comment]
After generating a new pdf with `compress: false`, is seems that the generated 'code' does treat it as any other character. As opposed to 'normal' spaces, that are removed from the output.

```
...
stream
0 Tr
0 0 0 rg
BT 56.69291 772.37068 Td /F6 12 Tf
0 Tc
0 Tw
100 Tz
0 Ts
[<00010002000300030004>]TJ ET
BT 88.7144 772.37068 Td /F6 12 Tf
0 Tc
0 Tw
100 Tz
0 Ts
[<0005>]TJ ET
BT 104.30034 772.37068 Td /F6 12 Tf
0 Tc
0 Tw
100 Tz
0 Ts
[<0005000600040007000300080009>]TJ ET
0 0 0 RG
0.6 w
...
10 beginbfchar
<0000> <0000>
<0001> <0048>
<0002> <0065>
<0003> <006C>
<0004> <006F>
<0005> <200B>
<0006> <0057>
<0007> <0072>
<0008> <0064>
<0009> <0021>
endbfchar
endcmap
...
```

**Comment by rspilker** [comment]
I've also tried several other white-space characters. None of them end up in the stream nor in the cmap.

If I recall correctly, from some unicode version on, the zero-width space stopped being a white-space character.

**Comment by rspilker** [comment]
So all other 'white-space characters' are either consumed in `RichText.layout()` by either `.split('\n')` to split the lines or `.split(RegExp(r'\s'))` to split the words.

In order to remove the zero-with space, which is not consumed (and rightfully so) by the word-splitter, there are three ppssible location with all their upsides and downsides.

1. The easiest it to just call `.replaceAll('\u200b', '')` early in the process, for instance just before the lines are split. 
Advantage: easy.
Downside: the hint for a possible word break point is lost. Currently not that big a downsize since there's no attempt to find a 'good' spot in `_splitWord` anyway.

2. Remove the character when the _Word object is created. Either in that constructor or where it's called.
Advantage: easy, allows for later improvements to the word-splitting algorithm.
Downside: the current word-splitting algorithm uses offsets and ensures at least one character is consumed. If that character happens to be a character that later is removed, I cannot easily prove invariants

3. Remove the character in `PdfGraphics.drawString`
Advantage: this problem is also solved for low-level PDF generation
Downside: It might be to low a level for still removing characters

I'm willing to create a pull request and add an automated test to verify no `\u200b` characters end up in the output. Please tell me if such a pull request would be accepted, and which of the three location you prefer (or possibly you can even come up with a fourth one).

**Comment by DavBfr** [comment]
Yes, I'll take a PR.
I think the second would be best.
Thanks

**Comment by rspilker** [comment]
First, I discovered that there are more 'formatting' hints that are not supported by the fonts. Also the `Word Joiner \u2060` is not handled properly. And oppositely, the `Soft Hyphen, &shy; \u00ad` is always rendered, while it should not be rendered unless it is used to break up a word at the end of a line.

Unfortunately, fixing it is not that easy.

In the function [`List<InlineSpan> _preprocessSpans(Context context)`](https://github.com/DavBfr/dart_pdf/blob/98d5c48b950b76580c1515077a5a8e24a6376b3a/pdf/lib/src/widgets/text.dart#L752), it is checked if the font supports all runes. Unfortunately, that's not to case. I can short-cut the code there, and allow several formatting codes to pass through, but then the next bump is [where the metrics are calculated](https://github.com/DavBfr/dart_pdf/blob/98d5c48b950b76580c1515077a5a8e24a6376b3a/pdf/lib/src/widgets/text.dart#L917).

Last, I noticed that all 'white space' is used as a word separator, and consumed in the process. This means that the widths and behaviors of those characters are ignored. In most cases not a problem, but it might be, for instance when aligning numbers in a table.

I think that the best solution is to introduce a new type of `InlineSpan`, possibly `_FormatSpan` and have the `List<InlineSpan> _preprocessSpans(Context context)` insert those in the proper locations. This can include new line, vertical tab, etc.

The the layout function no longer has to do it's own _parsing_ by splitting the text into lines or the lines into words.

This would be a bigger change than simply calling `text.replaceAll('[\u200b\u2060]', '')` in one spot.

Any thoughts and feedback welcome.
